feat(docs): add workflow to generate documentation on PR merge

virajbhartiya opened the pull request:
## Related Issues
Closes #12233 

## Proposed Changes
Add a workflow that run `make docsgen-cli` whenever the PR is merged

## Checklist

- [ ] Commits have a clear commit message.
- [ ] PR title conforms with [contribution conventions](https://github.com/filecoin-project/lotus/blob/master/CONTRIBUTING.md#pr-title-conventions)
- [ ] Update CHANGELOG.md or signal that this change does not need it per [contribution conventions](https://github.com/filecoin-project/lotus/blob/master/CONTRIBUTING.md#changelog-management)
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

rvagg commented on .github/workflows/docsgen-cli.yml:
```diff
@@ -0,0 +1,31 @@
+name: Generate Documentation
+
```
```suggestion

permissions:
  contents: write

```

I don't recall what our org settings are but this may be needed to be able to push changes from an action; doesn't hurt to have it either way

rvagg commented on .github/workflows/check.yml:
```diff
@@ -26,21 +26,19 @@ jobs:
       - uses: actions/checkout@v4
         with:
-          submodules: 'recursive'
+          submodules: "recursive"
```
you should check your editor and remove these auto-fix rules that you have, it's generally not good manners to go editing things unrelated to what you're intending to change without good reason to do so

virajbhartiya replied:
For some reason, my linter was automatically replacing the quotes, fixed it

rvagg commented:
@virajbhartiya can you resolve the conflict with `master` here please? You'll see that your `"` substitutions are getting in the way, so resolving those would be good.

virajbhartiya commented:
Hey @rvagg, I have resolved the conflicts in the PR

virajbhartiya commented:
@rvagg any suggestions on why is the workflow failing?

rvagg commented:
Failing because you need to install system dependencies for the build

in check.yml, you'll see:

```yml
      - uses: ./.github/actions/install-system-dependencies
      - uses: ./.github/actions/install-go
      - uses: ./.github/actions/make-deps
```

these are required to perform the actions you need, so you'll have to do it again in your new job

but, having said that, it's not cheap, so maybe this should all be in a single job so that's all done once and it's just the docsgen-cli bit that's done separately.

So how about this:

1. rename check.yml to check-and-gen.yml
2. add a `docsgen-cli` section to it with your new stuff, so that happens in the same flow

I'm still not convinced this is a great idea, it might get in the way. So we'll need to have a collective chat with others before we proceed.

virajbhartiya commented:
Hey @rvagg thanks for the review, I have currently modified it accordingly, renamed check.yml to check-and-gen.yml and added it in the same file. We can discuss regarding this on slack wtih other how to go ahead with this PR

BigLep commented on .github/workflows/check-and-gen.yml:
```diff
@@ -70,3 +68,37 @@ jobs:
+        run: |
+          if [ -n "$(git status --porcelain)" ]; then
+            git config user.name "github-actions[bot]"
+            git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
```
I have no expertise here, but maybe add a comment on where this comes from?  

Is https://github.com/orgs/community/discussions/26560 the canonical link?

virajbhartiya replied:
Yeah that's correct, I'll add a comment pointing to the GitHub discussion for future references

BigLep commented on .github/workflows/check-and-gen.yml:
```diff
@@ -70,3 +68,37 @@ jobs:
+            git config user.name "github-actions[bot]"
+            git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+            git add .
+            git commit -m "docs: update API documentation via docsgen-cli"
```
If we're adding a commit, I think it would be good to tie what PR this commit is asociated with. 
(We haven't developed team standards yet on whether commits should have a PR reference, but I think this is coming and we may as well do it here.)
Maybe @rvagg has input?
(This isn't blocking though.)

BigLep replied:
This can be disregarded.  I understand that this commit will be added before the PR is merged, and since  most PRs will be squashed, this goes away.  I'll resolve this.

BigLep commented:
I know there have been verbal conversation on this one.  Did we write out the next steps?

rvagg commented:
Didn't write it out, but next steps are to get this completed and merged and then experience it live and decide whether we hate it or love it and whether it's a QoL improvement on balance.

rvagg commented on .github/workflows/check-and-gen.yml:
```diff
@@ -70,3 +68,38 @@ jobs:
+  generate-docs:
+    name: Generate Documentation
+    runs-on: ubuntu-latest
+    if: github.event_name == 'push' && (github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/heads/release/'))
```
I'm unsure about this and whether a merge into one of these branches would trigger the job as well, we only want it to run on the PR branch, strictly. So this line worries me a little and it needs verification, somehow, or we need a subject-matter expert to help.

rvagg replied:
@BigLep maybe something to pull IPDX into to help review.

BigLep replied:
@galargh : do you know offhand whether this will only run on pushes to PR branches and not on master or release branches?

galargh replied:
That's the opposite of what this conditional says.

The conditional is: accept a workflow run which was triggered by a push event to either `master` or `release/*` branch.

rvagg commented on .github/workflows/check-and-gen.yml:
```diff
@@ -70,3 +68,38 @@ jobs:
+            git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+            git add .
+            git commit -m "docs: update API documentation via docsgen-cli"
+            git push
```
Another thing I'm wondering is if this is going to work to fork-branches, if this is run a a filecoin-project GHA runner, will we always have permissions to push back into the PR branch? Do we need to ensure we have a "can edit branches" thing turned on in our repo? Do we need to do anything else to make this work?

galargh replied:
As currently configured, this will be run on push events to `master` or `release/*`. But it won't work because they are protected. To overcome this, one would have to use a personal access token for a user that can disregard the branch protection rules.

rvagg commented on .github/workflows/check-and-gen.yml:
```diff
@@ -33,8 +33,6 @@ jobs:
       - uses: ./.github/actions/make-deps
       - run: make gen
```
Probably too big a leap for now, but I'd actually like to move `docsgen` out of `make gen` so we can run it here. Maybe for now we just do docsgen-cli and if it works OK we'll consider expanding it.

galargh reviewed (commented):
FYI, https://github.com/ipdxco/github-as-code/blob/master/.github/workflows/fix.yml is a working example of a workflow that runs on PRs and writes to PR branches.

BigLep commented:
@rvagg : do you think we can do the next steps in https://github.com/filecoin-project/lotus/pull/12681#issuecomment-2499426438 this week to run the experiment?  (Or have new thoughts come in and we defer for now.)

BigLep commented:
@virajbhartiya : is the final feedback something you'll be able to incorporate or does someone else need to take this?

virajbhartiya commented on .github/workflows/check-and-gen.yml:
```diff
@@ -70,3 +68,38 @@ jobs:
+  generate-docs:
+    name: Generate Documentation
+    runs-on: ubuntu-latest
+    if: github.event_name == 'push' && (github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/heads/release/'))
```
@BigLep So as for the changes what we need is to run on `pull_request` instead of the current condition so it runs only on PR branches or am I missing something?

BigLep replied:
@virajbhartiya : that's right.  We want this to run on the PR branches only.

rvagg commented:
I hope you don't mind @virajbhartiya but I've squashed this down to one commit, rebased on master, added a little fix to make sure this is only done on PRs and have force pushed all of that to your branch

rvagg commented:
https://github.com/filecoin-project/lotus/pull/12854

rvagg commented:
OK, I'm pushing my latest attempt at this and giving up for now, it's driving me mad.

* https://github.com/filecoin-project/lotus/pull/12856 is a PR testing it out, and that's just from _this_ repo, I gave up on testing it from a fork but that would need to be done too.
* It makes a merge commit, I think because actions/checkout does a detached head but I thought `ref:` was supposed to deal with that, argh!
* The push after commit wipes out all of the "Checks" so to find the actions runs you have to hunt through the commits and look for the ❌ or ✅ and get to it via that; this isn't OK and would be cause enough to bail on this whole project if this is what you have to put up with if an action pushes to a branch that it's checking.
* I really wanted to see a comment on the PR come out of it but didn't get past all the other frustration to make my commenter work, but that's probably not too hard, see error in https://github.com/filecoin-project/lotus/actions/runs/13024797399/job/36331973937?pr=12856

As of now, @virajbhartiya's `docsgen-workflow` branch has my latest. If this doesn't get picked up again soon I think we should just give up. `docsgen-cli` is much faster now than it once was and it may not be a bad thing to even bundle it with `docsgen`. Asking the user to do it is less onerous than it was when we started this.

BigLep commented:
@galargh : if you are up for timeboxing 1 hour to land this, that's great.   I wouldn't want to spend more than an hour here though as we are ok to close it too.

galargh commented:
> @galargh : if you are up for timeboxing 1 hour to land this, that's great. I wouldn't want to spend more than an hour here though as we are ok to close it too.

This is a minimal example of a workflow committing to pull requests (including public, non-org forks where the user allowed maintainer edits) - https://github.com/galorgh/.github/blob/master/.github/workflows/playground3.yml

It requires the use of `pull_request_target` to obtain credentials that are allowed writing to forks. One should be very careful when operating the the pull request target context because it has access to secrets and powerful credentials.

The pushed commit doesn't trigger any more workflows by design. This means it is not suitable for repos with required checks.

The way to overcome the limitations of this approach is to use custom credentials - either bot account's PAT or designated GItHub App's secrets.

rvagg commented:
> The pushed commit doesn't trigger any more workflows by design. This means it is not suitable for repos with required checks.

@galargh the problem I was having was that the pushed commit would make the previous workflows disappear in the original PR, making it pretty useless. It makes sense, because the checks were against the previous head but we just updated the head, but it's not the effect I was after.

Currently my vote is on bailing on this approach and just making the user deal with it. `make gen` has become faster to run, but I believe it still means you have to install dependencies and jump through all those hoops, which is unfortunate.

BigLep commented:
> Currently my vote is on bailing on this approach and just making the user deal with it. `make gen` has become faster to run, but I believe it still means you have to install dependencies and jump through all those hoops, which is unfortunate.

Ack.  I'm going to close this for now then.  This work was spawned from https://github.com/filecoin-project/lotus/issues/12233 which came from an operational event and this was identified as a way to potentially move quicker.  Per above, given the command has become faster, we at least moved in a better direction.
